src/v/io/interval_map.h

```diff
@@ -0,0 +1,211 @@
+     *
+     * If true is returned then the interval was inserted, and the corresponding
+     * iterator points to the inserted interval.
+     *
+     * If false is returned then the interval was not inserted. If insertion
+     * failed because the length was zero, then the returned iterator will be
+     * equal to end(). Otherwise, the iterator will point at an interval that
+     * overlapped with the interval being inserted.
+     *
+     * Invalidates iterators.
+     */
+    [[nodiscard]] std::pair<const_iterator, bool>
+    insert(T start, T length, V value);
+
+    /**
+     * Find the interval containing \p index.
+     *
+     * If no such interval exists then end() is returned.
+     */
+    [[nodiscard]] const_iterator find(T index) const;
+
+    /**
+     * Return an iterator to the first entry in the map.
+     *
+     * If the map is empty then end() is returned.
+     */
+    [[nodiscard]] const_iterator begin() const;
+
+    /**
+     * Return an iterator to the end of the map.
+     */
+    [[nodiscard]] const_iterator end() const;
+
+    /**
+     * Return true if the map contains no intervals.
+     */
+    [[nodiscard]] bool empty() const;
+
+    /**
+     * Erase the interval pointed to by the iterator \it.
+     *
+     * Invalidates iterators.
+     */
+    void erase(const_iterator it);
+
+private:
+    map_type map_;
+};
+
+template<std::integral T, typename V>
+std::pair<typename interval_map<T, V>::const_iterator, bool>
+interval_map<T, V>::insert(T start, T length, V value) {
+    if (length <= 0) {
+        return {map_.cend(), false};
+    }
+
+    const auto end = start + length;
+
+    auto it = map_.lower_bound(start);
+    if (it == map_.end()) {
+        /*
+         * all intervals in the container have starting offsets that are less
+         * than the starting offset of the interval being inserted.
+         */
+        if (map_.empty()) {
+            return map_.try_emplace({start, end}, value);
+        }
+
+        // checks for overlap with the interval on the left
+        it = std::prev(it);
+        if (it->first.end > start) {
+            return {it, false};
+        }
+
+        return {map_.try_emplace(it, {start, end}, value), true};
+    }
+
+    // checks for overlap with the interval on the right
+    if (end > it->first.start) {
+        return {it, false};
+    }
+
+    // there are no intervals on the left
+    if (it == map_.begin()) {
+        return map_.try_emplace({start, end}, value);
+    }
+
+    // checks for overlap with the interval on the left
+    it = std::prev(it);
+    if (it->first.end > start) {
+        return {it, false};
+    }
+
+    return {map_.try_emplace(it, {start, end}, value), true};
+}
+
+template<std::integral T, typename V>
+interval_map<T, V>::const_iterator interval_map<T, V>::find(T index) const {
+    auto it = map_.lower_bound(index);
+    if (it == map_.cend()) {
+        if (map_.empty()) {
+            return map_.cend();
+        }
```

Comment:
nit: could do this before the lower bound

Reply:
thought about it, but then wouldn't we incur that check even if it wasn't needed (e.g. it != end())? not that calling empty() has a cost that we should even remotely care about!